cloudinit/sources/DataSourceEc2.py

```diff
@@ -731,28 +735,91 @@ def convert_ec2_metadata_network_config(network_md, macs_to_nics=None,
        not provided, get_interfaces_by_mac is called to get it from the OS.
     @param: fallback_nic: Optionally provide the primary nic interface name.
        This nic will be guaranteed to minimally have a dhcp4 configuration.
+    @param: apply_network_config: Boolean set True to configure all networking
+       presented by IMDS. This includes rendering secondary IPv4 or IPv6
+       addresses on any NICs and rendering network config on secondary NICs.
 
-    @return A dict of network config version 1 based on the metadata and macs.
+    @return A dict of network config version 2 based on the metadata and macs.
     """
-    netcfg = {'version': 1, 'config': []}
+    netcfg = {'version': 2, 'ethernets': {}}
     if not macs_to_nics:
         macs_to_nics = net.get_interfaces_by_mac()
     macs_metadata = network_md['interfaces']['macs']
-    for mac, nic_name in macs_to_nics.items():
+    nic_idx = 1
+    for mac, nic_name in sorted(macs_to_nics.items()):
+        if not apply_network_config:
+            # Skip network config on all non-primary nics when
+            # apply_network_config is False (Xenial, Bionic, Eoan)
+            if fallback_nic and fallback_nic != nic_name:
+                continue
         nic_metadata = macs_metadata.get(mac)
         if not nic_metadata:
             continue  # Not a physical nic represented in metadata
-        nic_cfg = {'type': 'physical', 'name': nic_name, 'subnets': []}
-        nic_cfg['mac_address'] = mac
-        if (nic_name == fallback_nic or nic_metadata.get('public-ipv4s') or
-                nic_metadata.get('local-ipv4s')):
-            nic_cfg['subnets'].append({'type': 'dhcp4'})
-        if nic_metadata.get('ipv6s'):
-            nic_cfg['subnets'].append({'type': 'dhcp6'})
-        netcfg['config'].append(nic_cfg)
+        dhcp_override = {'route-metric': nic_idx * 100}
+        nic_idx += 1
+        dev_config = {'dhcp4': True, 'dhcp4-overrides': dhcp_override,
+                      'dhcp6': False,
+                      'match': {'macaddress': mac.lower()},
+                      'set-name': nic_name}
+        if nic_metadata.get('ipv6s'):  # Any IPv6 addresses configured
+            dev_config['dhcp6'] = True
+            dev_config['dhcp6-overrides'] = dhcp_override
+        if apply_network_config:
+            dev_config['addresses'] = get_secondary_addresses(nic_metadata)
+            if not dev_config['addresses']:
+                dev_config.pop('addresses')  # Since we found none configured
+        else:
+            LOG.debug(
+                'Skipping secondary IP address setup.'
+                ' Datasource configure_secondary_ips is False')
```

Comment:
And this else clause is not needed or at least should be moved up to the continue under the if not apply_network_config branch.

Reply:
done, moved it up and corrected it to mention apply_network_config and that secondary nics are ignored too.